python/lsst/scarlet/lite/image.py

```diff
@@ -0,0 +1,1372 @@
+        """Take the bitwise and of other and this."""
+        if type(other) in ScalarTypes:
+            return self.copy_with(data=other & self.data)
+        return cast(Image, other).__and__(self)
+
+    def __or__(self, other: Image | ScalarLike) -> Image:
+        """Take the binary or of this or other."""
+        return _operate_on_images(self, other, operator.or_)
+
+    def __ior__(self, other: Image | ScalarLike) -> Image:
+        """Take the binary or of this or other in place."""
+        return self._i_update(self.__or__, other)
+
+    def __ror__(self, other: Image | ScalarLike) -> Image:
+        """Take the binary or of other or this."""
+        if type(other) in ScalarTypes:
+            return self.copy_with(data=other | self.data)
+        return cast(Image, other).__or__(self)
+
+    def __xor__(self, other: Image | ScalarLike) -> Image:
+        """Take the binary xor of this xor other."""
+        return _operate_on_images(self, other, operator.xor)
+
+    def __ixor__(self, other: Image | ScalarLike) -> Image:
+        """Take the binary xor of this xor other in place."""
+        return self._i_update(self.__xor__, other)
+
+    def __rxor__(self, other: Image | ScalarLike) -> Image:
+        """Take the binary xor of other xor this."""
+        if type(other) in ScalarTypes:
+            return self.copy_with(data=other ^ self.data)
+        return cast(Image, other).__xor__(self)
+
+    def __lshift__(self, other: ScalarLike) -> Image:
+        """Shift this image to the left by other bits."""
+        if not issubclass(np.dtype(type(other)).type, np.integer):
+            raise TypeError("Bit shifting an image can only be done with integers")
+        return self.copy_with(data=self.data << other)
+
+    def __ilshift__(self, other: ScalarLike) -> Image:
+        """Shift this image to the left by other bits in place."""
+        self[:] = self.__lshift__(other)
+        return self
+
+    def __rlshift__(self, other: ScalarLike) -> Image:
+        """Shift other to the left by this image bits."""
+        return self.copy_with(data=other << self.data)
+
+    def __rshift__(self, other: ScalarLike) -> Image:
+        """Shift this image to the right by other bits."""
+        if not issubclass(np.dtype(type(other)).type, np.integer):
+            raise TypeError("Bit shifting an image can only be done with integers")
+        return self.copy_with(data=self.data >> other)
+
+    def __irshift__(self, other: ScalarLike) -> Image:
+        """Shift this image to the right by other bits in place."""
+        self[:] = self.__rshift__(other)
+        return self
+
+    def __rrshift__(self, other: ScalarLike) -> Image:
+        """Shift other to the right by this image bits."""
+        return self.copy_with(data=other >> self.data)
+
+    def __matmul__(self, other: Image | ScalarLike) -> Image:
+        """Matrix multiplication is not supported using images."""
```

Comment:
You're right. Originally this inherited from `numpy.ndarray`, so I had to overwrite `__matmul__` or unexpected things could happen. But I decided to not allow this to be used as a numpy array because (for example) adding an `image` with a `yx0` and a regular numpy array seemed like a footgun.